cartofacade/cgo_api.go

```diff
@@ -0,0 +1,401 @@
+
+	#include "../viam-cartographer/src/carto_facade/carto_facade.h"
+
+	viam_carto_lib* alloc_viam_carto_lib() { return (viam_carto_lib*) malloc(sizeof(viam_carto_lib)); }
+	void free_alloc_viam_carto_lib(viam_carto_lib* p) { free(p); }
+
+	viam_carto* alloc_viam_carto() { return (viam_carto*) malloc(sizeof(viam_carto)); }
+	void free_alloc_viam_carto(viam_carto* p) { free(p); }
+
+	bstring* alloc_bstring_array(size_t len) { return (bstring*) malloc(len * sizeof(bstring)); }
+	void free_bstring_array(bstring* p) { free(p); }
+*/
+import "C"
+
+import (
+	"context"
+	"errors"
+	"time"
+	"unsafe"
+)
+
+// CartoLib holds the c type viam_carto_lib
+type CartoLib struct {
+	value *C.viam_carto_lib
+}
+
+// Carto holds the c type viam_carto
+type Carto struct {
+	value *C.viam_carto
+}
+
+// AddSensorReading holds the c type viam_carto_sensor_reading
+type AddSensorReading struct {
+	value C.viam_carto_sensor_reading
+}
+
+// GetPositionResponse holds the c type viam_carto_get_position_response
+type GetPositionResponse struct {
+	value C.viam_carto_get_position_response
+}
+
+// GetPositionHolder holds values returned from c to be processed later
+type GetPositionHolder struct {
+	x float64
+	y float64
+	z float64
+
+	ox    float64
+	oy    float64
+	oz    float64
+	theta float64
+
+	real float64
+	imag float64
+	jmag float64
+	kmag float64
+
+	compReference string
+}
+
+// GetInternalStateResponse holds the c type viam_carto_get_internal_state_response
+type GetInternalStateResponse struct {
+	value C.viam_carto_get_internal_state_response
+}
+
+// LidarConfig represents the lidar configuration
+type LidarConfig int64
+
+const (
+	twoD LidarConfig = iota
+	threeD
+)
+
+// CartoConfig contains config values from app
+type CartoConfig struct {
+	sensors            []string
```

Comment:
Same question regarding whether these fields need to be public for the caller to set them

Reply:
yes, updating